sr/comp/mixtape/cli.py

```diff
@@ -202,6 +202,10 @@ def main():
         parser.print_help()
         return
 
+    if os.path.isfile(args.mixtape_directory):
+        print("File entered for the playlist when directory expected!")
+        return
```

Comment:
:+1: to the message, however as this is an error it would be great if we could exit with an error status code. That might mean we raise an exception (and ideally catch it in the next layer up) or perhaps just `exit` with a suitable code.

While we could do this ourselves, another approach might be to push this into the argument parsing (which would handle most of that for us). Suggest having a look at https://docs.python.org/3/library/argparse.html#type for how we can add custom validation. (Though let's agree a general path forwards for the CLI before worrying about implementing anything different).

Reply:
Apologies if this is a dumb question 
>  (and ideally catch it in the next layer up) 

What would be the next layer up in this case? Since this _is_ the main function?

Reply:
I am a silly goose - we have ```__main__.py``` so I guess in there?

Reply:
@PeterJCLaw something like 350607d? :)

Reply:
Perhaps, yeah. I'd probably let the callsite set the exception message and then just print it, rather than duplicating the message.

However:
- we still need to make the program exit with an error status code so it's clear this was an error rather than a happy exit
- I'd probably also send the error message to "standard error" rather than "standard output", for the same reason. This is a different output stream which is relevant mostly when you want to be able to direct the "real" output from a script somewhere different to control/logging messages, though it's good practice to do it all the time. Unless you do otherwise both streams show up in your terminal, so you probably wouldn't notice a difference.
- we probably don't need the custom exception type to be quite so specific (and maybe don't need a specific exception type at all, depending how we're structuring things)
- as I noted, really this error is part of parsing the input arguments -- so it would be great to push it to more clearly being part of that logic rather than a separate part (though this nice-to-have)

The simplest change from where you were might have been something like:
``` diff
@@ -203,8 +203,7 @@ def main():
         return
 
     if os.path.isfile(args.mixtape_directory):
-        print("File entered for the playlist when directory expected!")
-        return
+        exit("File entered for the playlist when directory expected!")
 
     if args.command == 'play':
         play(args)
```

This makes use of the built-in [`exit`](https://docs.python.org/3/library/sys.html#sys.exit) function which handles writing the message to `stderr` (rather than `stdout`) as well as quitting the program with an error status code. It's nicer to avoid calling `exit` anywhere other than the top level entry point, however since this function is already doing argument parsing (via `get_parser` and `parse_args`), which may call `exit` themselves, and the caller in `__main__.py` is raw module code (which I hadn't realised), this is probably fine and the easiest solution.

If there were a layer above I'd probably suggest introducing a custom but general exception (`CommandError` is  common name for this) which represents some error in the main level of the program and which means we need to error exit. That would be used pretty much like your `IncorrectDirectoryArgumentException` aside from the print in the caller using the exception's message rather than a literal and also being somewhat more re-usable.

Usage of that might look like:
``` python
try:
    cli.main()
except cli.CommandError as e:
    exit(e)
```

Though given the usage here, just calling `exit` inline from `main` is simpler and basically equivalent.

Aside: ideally the exit on line 201-203 should also exit with an error status code IMO, but let's leave that for now.

Reply:
I really appreciate your detailed feedback :) 

I've regressed the changes I did and put in the `exit` call in 9fdaa.

Reply:
No worries. Yup, that change is :+1: